### ISSUE TITLE
Update process results in merely uninstalling if internet connection interrupted

### ISSUE THREAD

**Issue by GoogleCodeExporter** [chrsmith/windows-package-manager#186]
```
What steps will reproduce the problem?
1. Start an update process.
2. Have an internet failure.
3. Observe as the program is uninstalled, new version is not installed, program 
no longer shows up in "installed" list of programs and must be located again 
manually in "not installed" category for manual re-installation.

What is the expected output? What do you see instead?
The update process should be:
1. Download new version.
2. Verified download is not corrupt.
3. Uninstall old version.
4. Install new version.
That way a connection issue would not result in the program being simply 
uninstalled.

What version of the product are you using? On what operating system?
Npackd 1.17.8
Win7x64
```

Original issue reported on code.google.com by `taltam...@gmail.com` on 7 Jan 2013 at 5:27

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Note: While the bug is that a failed connection causes uninstall rather then 
update. My means of discovering it (for the curious) is due to a server 
failure. My connection is just fine; but repeated attempts to download and 
install "PDF Creator" v1.6.1 have failed with the message:
Installing PDF Creator 1.6.1
Error 12029: A connection with the server could not be established

So clearly there is something wrong with the server. However the fact that such 
an issue would cause an uninstall show that the order of actions in the upgrade 
process is wrong, hence I could extrapolate the bug which I have listed above.
```

Original comment by `taltam...@gmail.com` on 7 Jan 2013 at 5:31

**Comment by GoogleCodeExporter** [comment]
```
the current order of actions was chosen due to inability of many programs to 
install different version in parallel
```

Original comment by `tim.lebe...@gmail.com` on 8 Jan 2013 at 3:42
* Changed state: **Thinking**
* Added labels: **Type-Enhancement**
* Removed labels: **Type-Defect**

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Installing two versions in parellel is NOT an issue with the order I am 
proposing.
Terms:
U = Uninstall old
D = Download
V = Verify Download
I = Install new

Current order: U-D-V-I
* Can result in the bug I described

Two versions in parellel required would be: D-V-I-U
* There is absolutely no reason to do this when you can instead do my suggested 
order

My suggested order: D-V-U-I
* Solves this bug, is the most sensible thing to do, and does not require 2 
installations to exist simultanously.
```

Original comment by `taltam...@gmail.com` on 8 Jan 2013 at 5:19

**Comment by GoogleCodeExporter** [comment]
```
so, its been 2 months. Has anyone seen my reply explaining why the order of 
operations I proposed solves the bug while ALSO avoiding the bugs of installing 
2 instances in parallel?
```

Original comment by `taltam...@gmail.com` on 5 Mar 2013 at 7:58

**Comment by GoogleCodeExporter** [comment]
```
I actually get an email for every comment. So the answer is yes, your 
suggestion is understood. I am not yet sure whether this should be implemented 
and if yes, in what time frame.
```

Original comment by `tim.lebe...@gmail.com` on 5 Mar 2013 at 7:48

**Comment by GoogleCodeExporter** [comment]
```
what do you mean by frame?
And why not implement this? it solves a bug and does not actually cause the 
parallel install bug. Is there another specific issue that you can see about 
this implementation method?
```

Original comment by `taltam...@gmail.com` on 5 Mar 2013 at 8:54

**Comment by GoogleCodeExporter** [comment]
```
> what do you mean by frame?
by "time frame" I mean "when", e.g. in 1.18, in 1.19, ...

> And why not implement this? 
I am not yet sure how this should be implemented, how complicated this is and 
whether it should be implemented now (e.g. it is the most important thing)

> it solves a bug and does not actually cause the parallel install bug. 
I don't see this as a bug. The program works as designed.

> Is there another specific issue that you can see about this implementation 
method?
no

```

Original comment by `tim.lebe...@gmail.com` on 6 Mar 2013 at 9:39

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
"I don't see this as a bug. The program works as designed."
You don't see it as a bug that npacked uninstalls the program the user told it 
to upgrade instead of performing the actual upgrade as requested by the user?
```

Original comment by `taltam...@gmail.com` on 7 Mar 2013 at 6:30

**Comment by GoogleCodeExporter** [comment]
```
no. 

There are no real "upgrades" in Npackd. A package version is first uninstalled 
and the new one is installed.
```

Original comment by `tim.lebe...@gmail.com` on 7 Mar 2013 at 6:03

**Comment by GoogleCodeExporter** [comment]
```
There are 3 buttons labeled "install, uninstall, and update" respectively.
And of 4 filters one is "installed, updateable".

The fact that an update is facilitated by first uninstalling and then 
installing does not change anything. There is still an upgrade feature and said 
feature still fails in a way that results in an entirely different process 
being performed (an uninstall only). Something that is entirely preventable.
```

Original comment by `taltam...@gmail.com` on 7 Mar 2013 at 6:12

**Comment by GoogleCodeExporter** [comment]
```
Issue 231 has been merged into this issue.
```

Original comment by `tim.lebe...@gmail.com` on 1 May 2013 at 5:05

**Comment by GoogleCodeExporter** [comment]
```
It's just one thing: Download before uninstall.

Why? To minimize the time without the package, and to prevent being locked out 
by a failing internet connection, left without any package.
```

Original comment by `FUrba...@gmail.com` on 2 Jun 2014 at 10:28
